provisioner/provisioner.go

```diff
@@ -0,0 +1,150 @@
+package provisioner
+
+import (
+	"fmt"
+	"github.com/yard-turkey/lib-bucket-provisioner/pkg/apis"
+	"github.com/yard-turkey/lib-bucket-provisioner/pkg/apis/objectbucket.io/v1alpha1"
+	claimReconciler "github.com/yard-turkey/lib-bucket-provisioner/provisioner/claim-reconciler"
+	bucketReconciler "github.com/yard-turkey/lib-bucket-provisioner/provisioner/object-bucket-reconciler"
+	"k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/util/validation"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+	"k8s.io/client-go/rest"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/manager/signals"
+	"strings"
+	"time"
+)
+
+const (
+	ProvisionerNamePrefix    = "object-bucket-claims.lib-bucket-provisioner/"
+	DefaultRetryBaseInterval = time.Second * 10
+	DefaultRetryTimeout      = time.Second * 360
+	DefaultRetryBackOff      = 1
+	DefaultThreadiness       = 1
+)
+
+// provisionerController is the first iteration of our internal provisioning
+// controller.  The passed-in bucket provisioner, coded by the user of the
+// library, is stored for later Provision and Delete calls.
+type provisionerController struct {
+	manager     manager.Manager
+	provisioner Provisioner
+	threads     int
+}
+
+type ProvisionerOptions struct {
+	// Threadiness is the amount of Reconciler routines per Controller
+	Threadiness int
+
+	// ProvisionBaseInterval the initial time interval before retrying
+	ProvisionBaseInterval time.Duration
+
+	// ProvisionRetryTimeout the maximum amount of time to attempt bucket provisioning.
+	// Once reached, the claim key is dropped and re-queued
+	ProvisionRetryTimeout time.Duration
+	// ProvisionRetryBackoff the base interval multiplier, applied each iteration
+	ProvisionRetryBackoff int
+}
+
+// NewProvisioner should be called by importers of this library to
+// instantiate a new provisioning controller. This controller will
+// respond to Add / Update / Delete events by calling the passed-in
+// provisioner's Provisioner and Delete methods.
+func NewProvisioner(cfg *rest.Config, provisionerName string, provisioner Provisioner, kubeVersion string, options *ProvisionerOptions) (*provisionerController, error) {
+
+	var err error
+
+	if nameErrors := isValidProvisionerName(provisionerName, field.NewPath("provisioner")); len(nameErrors) > 0 {
+		return nil, fmt.Errorf("invalid provisioner name %q: %v", provisionerName, nameErrors)
+	}
+
+	if options.Threadiness < DefaultThreadiness {
+		options.Threadiness = DefaultThreadiness
+	}
+	if options.ProvisionBaseInterval < DefaultRetryBaseInterval {
+		options.ProvisionBaseInterval = DefaultRetryBaseInterval
+	}
+	if options.ProvisionRetryTimeout < DefaultRetryTimeout {
+		options.ProvisionRetryTimeout = DefaultRetryTimeout
+	}
+	if options.ProvisionRetryBackoff < DefaultRetryBackOff {
```

Comment:
are there any upper limit checks on these passed-in numbers?

Reply:
At this point, i'm erring on the side of caution on behalf of the caller.   I'm assuming the caller is aware of the capability of the hardware their running.